sources/web/datalab/metadata.ts

```diff
@@ -0,0 +1,162 @@
+      const accountJSON: any = {
+        aliases: ["default"],
+        email: metadata.creds.account,
+        scopes: [metadata.creds.scopes],
+      };
+      res.writeHead(200, { 'Content-Type': 'application/json' });
+      res.write(JSON.stringify(accountJSON));
+    } else if (urlpath == '/computeMetadata/v1/instance/service-accounts/default/email' ||
+               urlpath == '/computeMetadata/v1/instance/service-accounts/' + metadata.creds.account + '/email') {
+      res.writeHead(200, { 'Content-Type': 'application/text' });
+      res.write(metadata.creds.account);
+    } else if (urlpath == '/computeMetadata/v1/instance/service-accounts/default/scopes' ||
+               urlpath == '/computeMetadata/v1/instance/service-accounts/' + metadata.creds.account + '/scopes') {
+      res.writeHead(200, { 'Content-Type': 'application/text' });
+      res.write(metadata.creds.scopes);
+    } else if (urlpath == '/computeMetadata/v1/instance/service-accounts/default/token' ||
+               urlpath == '/computeMetadata/v1/instance/service-accounts/' + metadata.creds.account + '/token') {
+      var token: any = {
+        access_token: metadata.creds.access_token,
+        expires_in: metadata.creds.expires_in,
+        token_type: metadata.creds.token_type,
+      };
+      res.writeHead(200, { 'Content-Type': 'application/json' });
+      res.write(JSON.stringify(token));
+    } else {
+      res.writeHead(404);
+    }
+    res.end();
+  });
+  server.listen(port, metadata_host, 511, () => {
+    // The `gcloud` tool uses a file in its config directory named 'gce' to
+    // cache whether or not it should read credentials from the metadata server.
+    //
+    // Since our fake metadata server just connected, it's possible that an
+    // earlier invocation of `gcloud` could have written that file. To account
+    // for this, we overwrite the file with the value that indicates the tool
+    // should read from the metadata server.
+    const gceFile = '/content/datalab/.config/gce'
+    fs.writeFileSync(gceFile, "True");
+  });
+}
+
+/**
+ * Initializes the Jupyter server manager.
+ */
+export function init(settings: common.AppSettings): void {
+  if (process.env.DATALAB_FAKE_METADATA_SERVER != 'true') {
+    return;
+  }
+  launchFakeServer(metadata);
+}
+
+function parseCreds(request: http.ServerRequest, callback: Function): void {
+  var body : string = "";
+  request.on('data', function(chunk: string) { body += chunk; });
+  request.on('end', function() {
+    callback(JSON.parse(body));
+  });
+}
+
+/**
+ * Implements the token submission request handling.
+ * @param request the incoming token POST request.
+ * @param response the outgoing response.
+ */
+function requestHandler(request: http.ServerRequest, response: http.ServerResponse): void {
+  var requestUrl = url.parse(request.url);
+  var path = requestUrl.pathname;
+  if (request.url.indexOf('/api/creds') === 0 && 'POST' == request.method) {
```

Comment:
Would this also match "/api/creds_not_really_the_creds_url"?  Not sure if we care.

Reply:
Yes, it would have. I don't think we care, but there's no reason to allow that, so I've changed it so that will not match.